src/buildContext.ts

```diff
@@ -0,0 +1,120 @@
+/* eslint-disable max-len */
+import passport, { AuthenticateOptions } from 'passport';
+import express from 'express';
+import { ExecutionParams } from 'subscriptions-transport-ws';
+import { UserTemplate, AuthenticateReturn, IVerifyOptions } from './types';
+
+const promisifiedAuthentication = (
+  req: express.Request,
+  res: express.Response,
+  name: string,
+  options: AuthenticateOptions,
+) => {
+  const p = new Promise<AuthenticateReturn>((resolve, reject) => {
+    const done = (
+      err: Error | undefined,
+      user: UserTemplate | undefined,
+      info?: IVerifyOptions | undefined,
+    ) => {
+      if (err) reject(err);
+      else resolve({ user, info });
+    };
+
+    const authFn = passport.authenticate(name, options, done);
+    return authFn(req, res);
+  });
+
+  return p;
+};
+
+const promisifiedLogin = (
+  req: express.Request,
+  user: UserTemplate,
+  options: AuthenticateOptions,
+) => {
+  const p = new Promise<void>((resolve, reject) => {
+    const done = (err: Error | undefined) => {
+      if (err) reject(err);
+      else resolve();
+    };
+
+    req.login(user, options, done);
+  });
+
+  return p;
+};
+
+export interface Context {
+  isAuthenticated: () => boolean;
+  isUnauthenticated: () => boolean;
+  getUser: () => UserTemplate;
+  authenticate: (strategyName: string, options?: object) => Promise<AuthenticateReturn>;
+  login: (user: UserTemplate, options?: object) => Promise<void>;
+  logout: () => void;
+  req: express.Request;
+  res?: express.Response;
+}
+
+const buildCommonContext = (req: express.Request, additionalContext: {}): Context => ({
+  isAuthenticated: () => req.isAuthenticated(),
+  isUnauthenticated: () => req.isUnauthenticated(),
+  getUser: () => req.user,
+  authenticate: (strategyName: string) => {
+    throw new Error(`Authenticate (${strategyName}) not implemented for subscriptions`);
```

Comment:
Good idea!

Reply:
Thanks! I've been working with the package with unit tests and there are some minor things to the Readme that I think may help (I probably spent an hour before I got `apollo-server-testing` to work). Also, the `PassportRequest` should be a generic accepting a the User interface that is returned.